operator/pkg/controller/virtualip/controller.go

```diff
@@ -0,0 +1,474 @@
+	case *appsv1.DaemonSet:
+		return r.mutateDaemonSet(obj.(*appsv1.DaemonSet))
+	default:
+		return nil
+	}
+}
+
+func (r *VirtualIPReconciler) mutateConfigMap(obj *corev1.ConfigMap) error {
+	poolName := obj.GetLabels()[labelPoolName]
+	poolInfo := r.getPools()[poolName]
+	nodes := r.getNodeList(poolName)
+
+	current := &HLConfig{}
+	if content, ok := obj.Data[hlConfigMapKey]; ok {
+		if err := current.Load(content); err != nil {
+			return err
+		}
+	}
+
+	desired := &HLConfig{}
+	desired.Init()
+	for _, ip := range poolInfo.Addresses {
+		addr := current.GetAddr(string(ip))
+		if addr == nil {
+			addr = &VIPAddress{
+				IP:   string(ip),
+				VrId: r.getFreeVRID(),
+			}
+			if addr.VrId == -1 {
+				return fmt.Errorf(
+					"unable to find any free Virtual Router ID for '%s' in pool '%s'",
+					ip, poolName,
+				)
+			}
+		}
+		if addr.Node != "" {
+			// The node is already defined
+			if nodes[addr.Node] == 0 {
+				// This node is not available, remove it
+				addr.Node = ""
+			} else {
+				nodes[addr.Node]--
+			}
+		}
+
+		desired.Addresses = append(desired.Addresses, *addr)
+	}
+
+	// Spread the remaining nodes
+	for node, nb := range nodes {
+		for i := 0; i < nb; i++ {
+			for index := range desired.Addresses {
+				if desired.Addresses[index].Node == "" {
+					desired.Addresses[index].Node = node
+					break
+				}
+			}
+		}
+	}
+
+	content, err := desired.ToYaml()
+	if err != nil {
+		return err
+	}
+	if obj.Data == nil {
+		obj.Data = make(map[string]string)
+	}
+	obj.Data[hlConfigMapKey] = content
+
+	// Store Config checksum
+	checksum := sha256.Sum256([]byte(content))
+	r.configChecksum[poolName] = hex.EncodeToString(checksum[:])
+
+	return nil
+}
+
+func (r *VirtualIPReconciler) mutateDaemonSet(obj *appsv1.DaemonSet) error {
+	poolName := obj.GetLabels()[labelPoolName]
+	poolInfo := r.getPools()[poolName]
+
+	selector := make(client.MatchingLabels)
+	selector[labelPoolName] = poolName
+
+	obj.Spec.Selector = &metav1.LabelSelector{
+		MatchLabels: selector,
+	}
+	utils.UpdateLabels(&obj.Spec.Template.ObjectMeta, selector)
+
+	obj.Spec.Template.Spec.NodeSelector = poolInfo.NodeSelector
+
+	utils.UpdateAnnotations(
+		&obj.Spec.Template.ObjectMeta,
+		map[string]string{annotationChecksumName: r.configChecksum[poolName]},
+	)
+
+	obj.Spec.Template.Spec.HostNetwork = true
+
+	volumeName := "hl-config"
+	defaultMode := int32(420)
+	obj.Spec.Template.Spec.Volumes = []corev1.Volume{
+		{
+			Name: volumeName,
+			VolumeSource: corev1.VolumeSource{
+				ConfigMap: &corev1.ConfigMapVolumeSource{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: r.getObjName(poolName),
+					},
+					DefaultMode: &defaultMode,
+				},
+			},
+		},
+	}
+
+	// If we have more than 1 container we clean up the containers list
+	if len(obj.Spec.Template.Spec.Containers) != 1 {
+		obj.Spec.Template.Spec.Containers = []corev1.Container{{}}
+	}
+	container := &obj.Spec.Template.Spec.Containers[0]
+
+	container.Name = "keepalived"
```

Comment:
I'm wondering wether we should abstract away this controller from keepalived and not take assumption that we are relying on it for the implementation. We could introduce an intermediate layer "VirtualIpProvider" or something similar which could then be implemented using keepalived, metallb or whatever else if a new virtualip implementation suits better our needs later on, this would ease keepalived replacement

Reply:
Hmm yes, we can do this, not sure it's really worth but ... may make sense